dashboard/app/models/script_level.rb

```diff
@@ -542,4 +546,9 @@ def hidden_for_section?(section_id)
   def should_hide_survey(user, viewed_user)
     anonymous? && user.try(:teacher?) && !viewed_user.nil? && user != viewed_user
   end
+
+  def self.generate_seeding_id(levels, lesson, lesson_group, script)
+    values = [lesson.key, lesson_group.key, script.name] + levels.map(&:unique_key)
+    HashingUtils.hash_values(values)
+  end
```

Comment:
I don't think I fully understand what ScriptLevels are supposed to be. It kind of feels like we have two levels of join tables between Script and Level (`script_levels` and `levels_script_levels`). Maybe if we finish modeling everything that we're planning to - with Activity and ActivitySection/Progression, etc., and deprecate variants - we no longer need to have both?

In general, if we want to simplify this more, we could add a globally unique seeding ID to each object that doesn't already have one. Even without changing what's stored in the DB, we can probably refactor the existing code such that every model implements a method like `generate_seeding_id` or maybe `seeding_equals`, which probably lets us parameterize the seeding logic for each level. I think the final code could look fairly nice and be declarative to add new models to, I just haven't invested the time into the cleanup yet.

Reply:
### script levels

> I don't think I fully understand what ScriptLevels are supposed to be. It kind of feels like we have two levels of join tables between Script and Level (`script_levels` and `levels_script_levels`). Maybe if we finish modeling everything that we're planning to - with Activity and ActivitySection/Progression, etc., and deprecate variants - we no longer need to have both?

A ScriptLevel object represents the presence of a level in a script.

The ScriptLevel object was a simple join table until variants came along. that when the levels_script_levels table had to be created, so that multiple levels could be associated with a single ScriptLevel. At that point it became a bit more confusing what a ScriptLevel is. This is why I want so badly to get rid of variants, because it would let us get rid of levels_script_levels.

The other way we could get rid of levels_script_levels is to use the sublevels pattern to turn level-swapping into a kind of level which has sublevels. the "sublevels" metaphor is a bit of a loose fit for level-swapping. We decided not to do this initially because it would create a bunch more levels for levelbuilders to have to manage, but this may not be so bad because (1) level-swapping is infrequent and (2) once we are on the script edit gui we could hide the fact that level-swapping uses sublevels from the curriculum writer.

I don't believe the Activity / ActivitySection work is necessary for this, which from a data model hierarchy perspective has mostly to do with taking the `progression` field off of ScriptLevel and making it into a new tier above it in the hierarchy.

Today, the best explanation for ScriptLevel is that it represents a "progress bubble" in the UI. A progress bubble has a position in a lesson, but might show different levels to different users (as explained about Variants earlier). If "progress bubble" is too specific to the UI layer for your taste, then another idea is to think of ScriptLevel generally as a thing that is no longer a simple join table and has grown to be its own entity and would best be renamed to no longer be named like a join table.

Reply:
Thanks for the explanation.

> this seems to give parity with the existing implementation, which would destroy/recreate a script level if any of the levels inside it change.

Agreed, this sounds right. I'd also be up for adding a unique key to ScriptLevel - that would simplify a lot of the logic - but the current implementation seems to work without it.

Resolving this just to declutter the PR conversation a bit.